dali/ft/filecopy.cpp

```diff
@@ -293,10 +293,161 @@ bool FileTransferThread::catchReadBuffer(ISocket * socket, MemoryBuffer & msg, u
+        }
+        else // find 1st of type 'spray'
+        {
+            Owned<IPropertyTreeIterator> directIOServices = getGlobalConfigSP()->getElements("services[@type='spray']");
+            if (directIOServices->first())
+                serviceTree.set(&directIOServices->query());
+            else
+                throw makeStringException(0, "launchFtSlaveCmd: no 'spray' dafilesrv services found");
+        }
+        connectEP.set(serviceTree->queryProp("@name"), serviceTree->getPropInt("@port"));
+    }
+#endif
+
+    StringBuffer url;
+    ep.getUrlStr(url);
+
+    Owned<ISocket> socket;
+    MemoryBuffer msg;
+    msg.setEndian(__BIG_ENDIAN);
+    if (sprayer.useFtSlave)
+    {
+        StringBuffer tmp;
+        socket.setown(spawnRemoteChild(SPAWNdfu, sprayer.querySlaveExecutable(ep, tmp), connectEP, DAFT_VERSION, queryFtSlaveLogDir(), this, wuid));
+        if (!socket)
+            throwError1(DFTERR_FailedStartSlave, url.str());
+
+        prepareCmd(msg, 0);
+        if (!catchWriteBuffer(socket, msg))
+            throwError1(RFSERR_TimeoutWaitConnect, url.str());
+    }
+    else
+    {
+        setDafsEndpointPort(connectEP);
+        if (connectEP.isNull())
+            return false;
+        socket.setown(connectDafs(connectEP, 5000));
+        if (!socket)
+            throwError1(DFTERR_FailedStartSlave, url.str());
+        prepareCmd(msg, daFileSrvCommandVersion);
+        sendDaFsFtSlaveCmd(socket, msg);
+    }
+    bool done;
+    for (;;)
+    {
+        msg.clear();
+        if (!catchReadBuffer(socket, msg, FTTIME_PROGRESS))
+            throwError1(RFSERR_TimeoutWaitSlave, url.str());
+
+        msg.setEndian(__BIG_ENDIAN);
+        msg.read(done);
+        if (done)
+            break;
+
+        OutputProgress newProgress;
+        newProgress.deserializeCore(msg);
+        newProgress.deserializeExtra(msg, 1);
+        newProgress.deserializeExtra(msg, 2);
+        sprayer.updateProgress(newProgress);
+
+        LOG(MCdebugProgress(10000), job, "Update %s: %d %" I64F "d->%" I64F "d", url.str(), newProgress.whichPartition, newProgress.inputLength, newProgress.outputLength);
+        if (isAborting())
+        {
+            if (!sendRemoteAbort(socket))
+                throwError1(RFSERR_TimeoutWaitSlave, url.str());
+        }
+    }
+    msg.read(ok);
+    setErrorOwn(deserializeException(msg));
+
+    LOG(MCdebugProgressDetail, job, "Finished generating part %s [%p] ok(%d) error(%d)", url.str(), this, (int)ok, (int)(error!=NULL));
+
+    // if communicating with ftslave, the process has a final ack wait
+    if (sprayer.useFtSlave)
+    {
+        // ftslave sends back a final ack.
+        msg.clear().append(true);
+        catchWriteBuffer(socket, msg);
+        if (sprayer.options->getPropInt("@fail", 0)) // JCSMORE - legacy? not set anywhere afaics
```

Comment:
I suspect it is a debug option which was used to test recovery or something similar.